File: day11/solution.py
from collections import Counter


def blink(counts: Counter[int]) -> None:
    old_counts = list(counts.items())
    counts.clear()
    for stone, count in old_counts:
        if stone == 0:
            counts[1] += count
        elif len(str(stone)) & 1 == 0:
            s = str(stone)
            half = len(s) // 2
            counts[int(s[:half])] += count
            counts[int(s[half:])] += count
        else:
            counts[stone * 2024] += count

def read_input() -> str:
    import sys, os
    if not os.isatty(sys.stdin.fileno()):
        return sys.stdin.read()
    with open(sys.argv[1] if len(sys.argv) > 1 else "input.txt") as f:
        return f.read()


def solve(nums: list[int]) -> None:
    stone_counter = Counter(nums)
    for _ in range(25): blink(stone_counter)
    print("Part 1:", sum(stone_counter.values()))
    for _ in range(50): blink(stone_counter)
    print("Part 2:", sum(stone_counter.values()))


if __name__ == "__main__":
    solve([int(s) for s in read_input().split()])

File: day11/solution.rs
use std::io::prelude::*;
use std::collections::*;

type Stone = u64;
type Day11Input = Vec<Stone>;

trait ReadDay11Input {
    fn read_input(&mut self) -> std::io::Result<Day11Input>;
}

impl<T: Read> ReadDay11Input for T {
    fn read_input(&mut self) -> std::io::Result<Day11Input> {
        let mut buf = String::new();
        self.read_to_string(&mut buf)?;
        let nums: Result<Day11Input, _> = buf.split_whitespace().map(str::parse).collect();
        Ok(nums.expect("input should be space-separated list of unsigned integers"))
    }
}

fn solve(nums: &Day11Input, blinks: usize) -> usize {
    let mut counts: HashMap<Stone, usize> = HashMap::new();
    for num in nums {
        *counts.entry(*num).or_insert(0) += 1;
    }
    const TEN: Stone = 10;
    for _ in 0..blinks {
        let old_counts = counts.iter().map(|(k, v)| (*k, *v)).collect::<Vec<_>>();
        counts.clear();
        for (k, count) in old_counts {
            // If the stone is engraved with the number 0, it is
            // replaced by a stone engraved with the number 1.
            if k == 0 {
                *counts.entry(1).or_insert(0) += count;
                continue
            }

            // If the stone is engraved with a number that has an
            // even number of digits, it is replaced by two stones
            let num_digits = k.ilog10() + 1;
            if num_digits & 1 == 0 {
                let pow = TEN.pow(num_digits / 2);
                let left = k / pow;
                let right = k.rem_euclid(pow);
                *counts.entry(left).or_insert(0) += count;
                *counts.entry(right).or_insert(0) += count;
                continue
            }

            // Otherwise, the stone is replaced by a new stone; the new
            // stone's number is old stone's number multiplied by 2024
            let num = 2024 * k;
            *counts.entry(num).or_insert(0) += count;
        }
    }
    counts.values().sum()
}

fn main() -> std::io::Result<()> {
    let input = match std::env::args().nth(1) {
        Some(ref value) if value == "--example" => vec![125, 17],
        Some(fname) => std::fs::File::open(fname)?.read_input()?,
        None => std::io::stdin().read_input()?,
    };
    println!("input = {:?}", input);
    println!("Part 1: {}", solve(&input, 25));
    println!("Part 2: {}", solve(&input, 75));
    Ok(())
}